kotlinx-coroutines-core/common/src/channels/Produce.kt

```diff
@@ -58,35 +76,187 @@ public suspend fun ProducerScope<*>.awaitClose(block: () -> Unit = {}) {
+ *
+ * ```
+ * val produceJob = Job()
+ * // create and populate a channel with a buffer
+ * val channel = produce<Int>(produceJob, capacity = Channel.UNLIMITED) {
+ *     repeat(5) { send(it) }
+ *     throw TestException()
+ * }
+ * produceJob.join() // wait for `produce` to fail
+ * check(produceJob.isCancelled == true)
+ * // prints 0, 1, 2, 3, 4, then throws `TestException`
+ * for (value in channel) { println(value) }
+ * ```
+ *
+ * When the coroutine is cancelled via structured concurrency and not the `cancel` function,
+ * the channel does not automatically close until the coroutine completes,
+ * so it is possible that some elements will be sent even after the coroutine is cancelled:
+ *
+ * ```
+ * val parentScope = CoroutineScope(Dispatchers.Default)
+ * val channel = parentScope.produce<Int>(capacity = Channel.UNLIMITED) {
+ *     repeat(5) {
+ *         send(it)
+ *     }
+ *     parentScope.cancel()
+ *     // suspending after this point would fail, but sending succeeds
+ *     send(-1)
+ * }
+ * for (c in channel) {
+ *     println(c) // 0, 1, 2, 3, 4, -1
+ * } // throws a `CancellationException` exception after reaching -1
+ * ```
+ *
+ * Note that cancelling `produce` via structured concurrency closes the channel with a cause,
+ * making it a _failed_ channel.
+ *
+ * The behavior around coroutine cancellation and error handling is experimental and may change in a future release.
+ *
+ * ### Coroutine context
  *
  * The coroutine context is inherited from this [CoroutineScope]. Additional context elements can be specified with the [context] argument.
  * If the context does not have any dispatcher or other [ContinuationInterceptor], then [Dispatchers.Default] is used.
  * The parent job is inherited from the [CoroutineScope] as well, but it can also be overridden
  * with a corresponding [context] element.
  *
- * Any uncaught exception in this coroutine will close the channel with this exception as the cause and
- * the resulting channel will become _failed_, so that any attempt to receive from it thereafter will throw an exception.
+ * See [newCoroutineContext] for a description of debugging facilities available for newly created coroutines.
  *
- * The kind of the resulting channel depends on the specified [capacity] parameter.
- * See the [Channel] interface documentation for details.
+ * ### Undelivered elements
  *
- * See [newCoroutineContext] for a description of debugging facilities available for newly created coroutines.
+ * Some values that [produce] creates may be lost:
+ *
+ * ```
+ * val channel = produce(Dispatchers.Default, capacity = 5) {
+ *     repeat(100) {
+ *         send(it)
+ *         println("Sent $it")
+ *     }
+ * }
+ * channel.cancel() // no elements can be received after this!
+ * ```
+ *
+ * There is no way to recover these lost elements.
+ * If this is unsuitable, please create a [Channel] manually and pass the `onUndeliveredElement` callback to the
+ * constructor.
```

Comment:
This example is irrelevant to `produce` documentation, it's `Channel` doc that should explain that

Reply:
I beg to differ. `produce` calls `Channel` function, passing a `null` to `onUndeliveredElement`. This is *`produce`'s* implementation detail (that we must document) and not the `Channel`'s.

Reply:
Or do you mean the snippet below? It's not an example, it's an explanation of how to approximate the `produce` above with an explicitly created channel. Would it work for you if I explain this in text explicitly?

Reply:
I mean the snippet itself, yes. I would say it's sufficient to refer to `onUndeliveredElement` where it has a more comprehensive explanation and self-sufficient example

Reply:
Non-obvious things this snippet highlights:

* Where the `onUndeliveredParameter` is (you can't refer to it directly, the user has to chase the docs)
* `finally` with the `close` in the coroutine

Reply:
>Where the onUndeliveredParameter is (you can't refer to it directly, the user has to chase the docs)

* "If this is unsuitable, please create a [Channel] manually and pass the `onUndeliveredElement` callback to the constructor: [Channel(onUndeliveredElement = ...)]]Channel]" probably will suffice.

I realize that using it might be less obvious, but still believe it's not the produce's documentation job to explain that